src/bio/bio_context.c

```diff
@@ -533,26 +593,75 @@ bio_ioctxt_open(struct bio_io_context **pctxt, struct bio_xs_context *xs_ctxt,
 	uuid_copy(ctxt->bic_pool_id, uuid);
 
 	/* NVMe isn't configured or pool doesn't have NVMe partition */
-	if (!bio_nvme_configured() || skip_blob) {
+	if (!bio_nvme_configured() || flags & BIO_MC_FL_NO_DATABLOB) {
 		*pctxt = ctxt;
 		return 0;
 	}
 
-	rc = bio_bs_hold(xs_ctxt->bxc_blobstore);
+	rc = bio_bs_hold(xs_ctxt->bxc_xs_blobstores[st]->bxb_blobstore);
 	if (rc) {
 		D_FREE(ctxt);
 		return rc;
 	}
 
+	ctxt->bic_xs_blobstore = xs_ctxt->bxc_xs_blobstores[st];
 	rc = bio_blob_open(ctxt, false);
 	if (rc) {
 		D_FREE(ctxt);
 	} else {
-		d_list_add_tail(&ctxt->bic_link, &xs_ctxt->bxc_io_ctxts);
+		d_list_add_tail(&ctxt->bic_link,
+				&xs_ctxt->bxc_xs_blobstores[st]->bxb_io_ctxts);
 		*pctxt = ctxt;
 	}
 
-	bio_bs_unhold(xs_ctxt->bxc_blobstore);
+	bio_bs_unhold(xs_ctxt->bxc_xs_blobstores[st]->bxb_blobstore);
+	return rc;
+}
+
+int bio_data_ioctxt_open(struct bio_io_context **pctxt, struct bio_xs_context *xs_ctxt,
+			 struct umem_instance *umem, uuid_t uuid)
+{
+	return bio_ioctxt_open(pctxt, xs_ctxt, umem, uuid, 0, SMD_TYPE_DATA);
+}
+
+int bio_mc_open(struct bio_xs_context *xs_ctxt, uuid_t pool_id, struct umem_instance *umm,
+		enum bio_mc_flags flags, struct bio_meta_context **mc)
+{
+	struct bio_meta_context	*bio_mc;
+	int			 rc;
+
+	D_ALLOC_PTR(bio_mc);
+	if (bio_mc == NULL)
+		return -DER_NOMEM;
+
+	rc = bio_ioctxt_open(&bio_mc->mc_data, xs_ctxt, umm, pool_id, flags, SMD_TYPE_DATA);
+	if (rc)
+		goto failed;
+
+	if (flags & BIO_MC_FL_NO_DATABLOB)
+		goto out;
+
+	if (xs_ctxt->bxc_xs_blobstores[SMD_TYPE_META]) {
+		rc = bio_ioctxt_open(&bio_mc->mc_meta, xs_ctxt, umm, pool_id,
+				     flags, SMD_TYPE_META);
+		if (rc)
+			goto failed;
+	}
+
+	if (xs_ctxt->bxc_xs_blobstores[SMD_TYPE_WAL]) {
+		rc = bio_ioctxt_open(&bio_mc->mc_wal, xs_ctxt, umm, pool_id,
+				     flags, SMD_TYPE_WAL);
+		if (rc)
+			goto failed;
+	}
```

Comment:
The mc_create/open/close/destroy() logic looks don't support blobstore sharing. (WAL, meta sharing same ssd, or WAL, meta & data sharing same SSD, etc.)

Reply:
yup, missed this.